File: Android/src/build-backups/kapt3-20260411-195127/stubs/debug/selfgemma/talk/feature/roleplay/chat/RoleplayChatViewModel.java
package selfgemma.talk.feature.roleplay.chat;

@kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000\u00b8\u0001\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010\u000e\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000b\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\t\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010\u0002\n\u0002\b\u0007\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\b\u0006\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u000b\b\u0007\u0018\u00002\u00020\u0001BK\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\b\b\u0001\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u0012\u0006\u0010\b\u001a\u00020\t\u0012\u0006\u0010\n\u001a\u00020\u000b\u0012\u0006\u0010\f\u001a\u00020\r\u0012\u0006\u0010\u000e\u001a\u00020\u000f\u0012\u0006\u0010\u0010\u001a\u00020\u0011\u00a2\u0006\u0004\b\u0012\u0010\u0013J\u000e\u0010.\u001a\u00020/2\u0006\u00100\u001a\u00020\u0015J\u000e\u00101\u001a\u00020/2\u0006\u00102\u001a\u00020!J\u000e\u00103\u001a\u00020/2\u0006\u00104\u001a\u00020\u0015J\u000e\u00105\u001a\u00020/2\u0006\u00106\u001a\u000207J\u0016\u00108\u001a\u00020/2\u0006\u00109\u001a\u00020\u00152\u0006\u0010:\u001a\u00020;J\u0010\u0010<\u001a\u00020/2\u0006\u0010=\u001a\u00020\u0015H\u0002J\u0010\u0010>\u001a\u00020/2\u0006\u00102\u001a\u00020!H\u0002J\u0010\u0010?\u001a\u00020/2\u0006\u00102\u001a\u00020!H\u0002J\b\u0010@\u001a\u00020/H\u0002J\u001c\u0010A\u001a\b\u0012\u0004\u0012\u00020C0B2\u0006\u0010D\u001a\u00020&H\u0082@\u00a2\u0006\u0002\u0010EJ\u0010\u0010F\u001a\u00020G2\u0006\u0010H\u001a\u00020\u0015H\u0002J\u001e\u0010I\u001a\u00020J2\f\u0010K\u001a\b\u0012\u0004\u0012\u0002070B2\u0006\u00102\u001a\u00020!H\u0002J\b\u0010L\u001a\u00020\u001fH\u0002J\u000e\u0010M\u001a\u00020/H\u0082@\u00a2\u0006\u0002\u0010NJ*\u0010O\u001a\b\u0012\u0004\u0012\u0002070B2\f\u0010P\u001a\b\u0012\u0004\u0012\u0002070B2\f\u0010Q\u001a\b\u0012\u0004\u0012\u00020G0BH\u0002J\f\u0010R\u001a\u00020\u0015*\u00020\u0015H\u0002J\b\u0010S\u001a\u00020/H\u0002J\b\u0010T\u001a\u00020/H\u0002R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\u000bX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\f\u001a\u00020\rX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000e\u001a\u00020\u000fX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0010\u001a\u00020\u0011X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0014\u001a\u00020\u0015X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0014\u0010\u0016\u001a\b\u0012\u0004\u0012\u00020\u00150\u0017X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0014\u0010\u0018\u001a\b\u0012\u0004\u0012\u00020\u00190\u0017X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0014\u0010\u001a\u001a\b\u0012\u0004\u0012\u00020\u001b0\u0017X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0010\u0010\u001c\u001a\u0004\u0018\u00010\u001dX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u001e\u001a\u00020\u001fX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0010\u0010 \u001a\u0004\u0018\u00010!X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0010\u0010\"\u001a\u0004\u0018\u00010\u0015X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010#\u001a\u00020\u001bX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0016\u0010$\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010&0%X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0016\u0010\'\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010(0%X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0017\u0010)\u001a\b\u0012\u0004\u0012\u00020+0*\u00a2\u0006\b\n\u0000\u001a\u0004\b,\u0010-\u00a8\u0006U"}, d2 = {"Lselfgemma/talk/feature/roleplay/chat/RoleplayChatViewModel;", "Landroidx/lifecycle/ViewModel;", "savedStateHandle", "Landroidx/lifecycle/SavedStateHandle;", "appContext", "Landroid/content/Context;", "dataStoreRepository", "Lselfgemma/talk/data/DataStoreRepository;", "conversationRepository", "Lselfgemma/talk/domain/roleplay/repository/ConversationRepository;", "roleRepository", "Lselfgemma/talk/domain/roleplay/repository/RoleRepository;", "memoryRepository", "Lselfgemma/talk/domain/roleplay/repository/MemoryRepository;", "sendRoleplayMessageUseCase", "Lselfgemma/talk/domain/roleplay/usecase/SendRoleplayMessageUseCase;", "extractMemoriesUseCase", "Lselfgemma/talk/domain/roleplay/usecase/ExtractMemoriesUseCase;", "<init>", "(Landroidx/lifecycle/SavedStateHandle;Landroid/content/Context;Lselfgemma/talk/data/DataStoreRepository;Lselfgemma/talk/domain/roleplay/repository/ConversationRepository;Lselfgemma/talk/domain/roleplay/repository/RoleRepository;Lselfgemma/talk/domain/roleplay/repository/MemoryRepository;Lselfgemma/talk/domain/roleplay/usecase/SendRoleplayMessageUseCase;Lselfgemma/talk/domain/roleplay/usecase/ExtractMemoriesUseCase;)V", "sessionId", "", "draft", "Lkotlinx/coroutines/flow/MutableStateFlow;", "metaState", "Lselfgemma/talk/feature/roleplay/chat/RoleplayChatMetaState;", "stopRequested", "", "dispatchJob", "Lkotlinx/coroutines/Job;", "lastDraftEditAtElapsed", "", "latestQueuedModel", "Lselfgemma/talk/data/Model;", "activeAssistantMessageId", "activeDispatchSuperseded", "sessionFlow", "Lkotlinx/coroutines/flow/Flow;", "Lselfgemma/talk/domain/roleplay/model/Session;", "roleFlow", "Lselfgemma/talk/domain/roleplay/model/RoleCard;", "uiState", "Lkotlinx/coroutines/flow/StateFlow;", "Lselfgemma/talk/feature/roleplay/chat/RoleplayChatUiState;", "getUiState", "()Lkotlinx/coroutines/flow/StateFlow;", "updateDraft", "", "value", "sendMessage", "model", "switchModel", "modelId", "pinMessage", "message", "Lselfgemma/talk/domain/roleplay/model/Message;", "addManualMemory", "content", "category", "Lselfgemma/talk/domain/roleplay/model/MemoryCategory;", "scheduleDispatch", "reason", "requestMergeAndStop", "dispatchPendingMessages", "refreshSupplementalState", "loadPinnedMemories", "", "Lselfgemma/talk/domain/roleplay/model/MemoryItem;", "session", "(Lselfgemma/talk/domain/roleplay/model/Session;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "stagePendingUserMessage", "Lselfgemma/talk/feature/roleplay/chat/QueuedUserMessage;", "input", "stageDispatchTurn", "Lselfgemma/talk/domain/roleplay/usecase/StagedRoleplayTurn;", "userMessages", "remainingDispatchDelay", "retractActiveAssistantBubble", "(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "mergeMessages", "messages", "queuedMessages", "escapeJson", "playSendSound", "playReceiveSound", "app_debug"})
@dagger.hilt.android.lifecycle.HiltViewModel()
public final class RoleplayChatViewModel extends androidx.lifecycle.ViewModel {
    @org.jetbrains.annotations.NotNull()
    private final android.content.Context appContext = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.data.DataStoreRepository dataStoreRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.repository.ConversationRepository conversationRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.repository.RoleRepository roleRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.repository.MemoryRepository memoryRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.SendRoleplayMessageUseCase sendRoleplayMessageUseCase = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.domain.roleplay.usecase.ExtractMemoriesUseCase extractMemoriesUseCase = null;
    @org.jetbrains.annotations.NotNull()
    private final java.lang.String sessionId = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.MutableStateFlow<java.lang.String> draft = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.MutableStateFlow<selfgemma.talk.feature.roleplay.chat.RoleplayChatMetaState> metaState = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.MutableStateFlow<java.lang.Boolean> stopRequested = null;
    @org.jetbrains.annotations.Nullable()
    private kotlinx.coroutines.Job dispatchJob;
    private long lastDraftEditAtElapsed = 0L;
    @org.jetbrains.annotations.Nullable()
    private selfgemma.talk.data.Model latestQueuedModel;
    @org.jetbrains.annotations.Nullable()
    private java.lang.String activeAssistantMessageId;
    private boolean activeDispatchSuperseded = false;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.Flow<selfgemma.talk.domain.roleplay.model.Session> sessionFlow = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.Flow<selfgemma.talk.domain.roleplay.model.RoleCard> roleFlow = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.StateFlow<selfgemma.talk.feature.roleplay.chat.RoleplayChatUiState> uiState = null;
    
    @javax.inject.Inject()
    public RoleplayChatViewModel(@org.jetbrains.annotations.NotNull()
    androidx.lifecycle.SavedStateHandle savedStateHandle, @dagger.hilt.android.qualifiers.ApplicationContext()
    @org.jetbrains.annotations.NotNull()
    android.content.Context appContext, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.data.DataStoreRepository dataStoreRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.repository.ConversationRepository conversationRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.repository.RoleRepository roleRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.repository.MemoryRepository memoryRepository, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.SendRoleplayMessageUseCase sendRoleplayMessageUseCase, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.usecase.ExtractMemoriesUseCase extractMemoriesUseCase) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.flow.StateFlow<selfgemma.talk.feature.roleplay.chat.RoleplayChatUiState> getUiState() {
        return null;
    }
    
    public final void updateDraft(@org.jetbrains.annotations.NotNull()
    java.lang.String value) {
    }
    
    public final void sendMessage(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.data.Model model) {
    }
    
    public final void switchModel(@org.jetbrains.annotations.NotNull()
    java.lang.String modelId) {
    }
    
    public final void pinMessage(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.Message message) {
    }
    
    public final void addManualMemory(@org.jetbrains.annotations.NotNull()
    java.lang.String content, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.MemoryCategory category) {
    }
    
    private final void scheduleDispatch(java.lang.String reason) {
    }
    
    private final void requestMergeAndStop(selfgemma.talk.data.Model model) {
    }
    
    private final void dispatchPendingMessages(selfgemma.talk.data.Model model) {
    }
    
    private final void refreshSupplementalState() {
    }
    
    private final java.lang.Object loadPinnedMemories(selfgemma.talk.domain.roleplay.model.Session session, kotlin.coroutines.Continuation<? super java.util.List<selfgemma.talk.domain.roleplay.model.MemoryItem>> $completion) {
        return null;
    }
    
    private final selfgemma.talk.feature.roleplay.chat.QueuedUserMessage stagePendingUserMessage(java.lang.String input) {
        return null;
    }
    
    private final selfgemma.talk.domain.roleplay.usecase.StagedRoleplayTurn stageDispatchTurn(java.util.List<selfgemma.talk.domain.roleplay.model.Message> userMessages, selfgemma.talk.data.Model model) {
        return null;
    }
    
    private final long remainingDispatchDelay() {
        return 0L;
    }
    
    private final java.lang.Object retractActiveAssistantBubble(kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    private final java.util.List<selfgemma.talk.domain.roleplay.model.Message> mergeMessages(java.util.List<selfgemma.talk.domain.roleplay.model.Message> messages, java.util.List<selfgemma.talk.feature.roleplay.chat.QueuedUserMessage> queuedMessages) {
        return null;
    }
    
    private final java.lang.String escapeJson(java.lang.String $this$escapeJson) {
        return null;
    }
    
    private final void playSendSound() {
    }
    
    private final void playReceiveSound() {
    }
}

File: Android/src/app/src/main/java/selfgemma/talk/feature/roleplay/chat/RoleplayChatViewModel.kt
package selfgemma.talk.feature.roleplay.chat

import android.content.Context
import android.os.SystemClock
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import java.util.UUID
import javax.inject.Inject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import selfgemma.talk.data.ConfigKeys
import selfgemma.talk.data.DataStoreRepository
import selfgemma.talk.data.Model
import selfgemma.talk.domain.roleplay.model.MemoryCategory
import selfgemma.talk.domain.roleplay.model.MemoryItem
import selfgemma.talk.domain.roleplay.model.Message
import selfgemma.talk.domain.roleplay.model.MessageSide
import selfgemma.talk.domain.roleplay.model.MessageStatus
import selfgemma.talk.domain.roleplay.model.RoleCard
import selfgemma.talk.domain.roleplay.model.Session
import selfgemma.talk.domain.roleplay.model.SessionEvent
import selfgemma.talk.domain.roleplay.model.SessionEventType
import selfgemma.talk.domain.roleplay.model.SessionSummary
import selfgemma.talk.domain.roleplay.repository.ConversationRepository
import selfgemma.talk.domain.roleplay.repository.MemoryRepository
import selfgemma.talk.domain.roleplay.repository.RoleRepository
import selfgemma.talk.domain.roleplay.usecase.ExtractMemoriesUseCase
import selfgemma.talk.domain.roleplay.usecase.SendRoleplayMessageUseCase
import selfgemma.talk.domain.roleplay.usecase.StagedRoleplayTurn
import selfgemma.talk.runtime.runtimeHelper

data class RoleplayChatUiState(
  val loading: Boolean = true,
  val session: Session? = null,
  val role: RoleCard? = null,
  val messages: List<Message> = emptyList(),
  val draft: String = "",
  val summary: SessionSummary? = null,
  val pinnedMemories: List<MemoryItem> = emptyList(),
  val inProgress: Boolean = false,
  val errorMessage: String? = null,
)

private const val TAG = "RoleplayChatViewModel"
private const val SEND_DISPATCH_DELAY_MS = 2_000L

private data class QueuedUserMessage(
  val message: Message,
  val persisted: Boolean = false,
)

private data class RoleplayChatMetaState(
  val summary: SessionSummary? = null,
  val pinnedMemories: List<MemoryItem> = emptyList(),
  val pendingUserMessages: List<QueuedUserMessage> = emptyList(),
  val inProgress: Boolean = false,
  val errorMessage: String? = null,
)

@HiltViewModel
class RoleplayChatViewModel
@Inject
constructor(
  savedStateHandle: SavedStateHandle,
  @ApplicationContext private val appContext: Context,
  private val dataStoreRepository: DataStoreRepository,
  private val conversationRepository: ConversationRepository,
  private val roleRepository: RoleRepository,
  private val memoryRepository: MemoryRepository,
  private val sendRoleplayMessageUseCase: SendRoleplayMessageUseCase,
  private val extractMemoriesUseCase: ExtractMemoriesUseCase,
) : ViewModel() {
  private val sessionId: String = checkNotNull(savedStateHandle["sessionId"])
  private val draft = MutableStateFlow("")
  private val metaState = MutableStateFlow(RoleplayChatMetaState())
  private val stopRequested = MutableStateFlow(false)
  private var dispatchJob: Job? = null
  private var lastDraftEditAtElapsed = 0L
  private var latestQueuedModel: Model? = null
  private var activeAssistantMessageId: String? = null
  private var activeDispatchSuperseded = false

  private val sessionFlow =
    conversationRepository.observeSessions().map { sessions ->
      sessions.firstOrNull { it.id == sessionId }
    }.distinctUntilChanged()
  private val roleFlow =
    combine(sessionFlow, roleRepository.observeRoles()) { session, roles ->
      roles.firstOrNull { it.id == session?.roleId }
    }.distinctUntilChanged()

  val uiState: StateFlow<RoleplayChatUiState> =
    combine(
      sessionFlow,
      conversationRepository.observeMessages(sessionId).distinctUntilChanged(),
      roleFlow,
      draft,
      metaState,
    ) { session, messages, role, draftValue, meta ->
      RoleplayChatUiState(
        loading = session == null,
        session = session,
        role = role,
        messages = mergeMessages(messages = messages, queuedMessages = meta.pendingUserMessages),
        draft = draftValue,
        summary = meta.summary,
        pinnedMemories = meta.pinnedMemories,
        inProgress = meta.inProgress,
        errorMessage = meta.errorMessage,
      )
    }
      .stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = RoleplayChatUiState(),
      )

  init {
    RoleplaySoundEffectPlayer.prepare(appContext)
    refreshSupplementalState()
  }

  fun updateDraft(value: String) {
    if (draft.value == value) {
      return
    }

    draft.value = value
    lastDraftEditAtElapsed = SystemClock.elapsedRealtime()
    if (metaState.value.pendingUserMessages.isNotEmpty() && !metaState.value.inProgress) {
      scheduleDispatch(reason = "draft changed while send pending")
    }
  }

  fun sendMessage(model: Model) {
    val input = draft.value.trim()
    if (input.isBlank()) {
      return
    }

    latestQueuedModel = model
    draft.value = ""
    lastDraftEditAtElapsed = SystemClock.elapsedRealtime()

    val queuedMessage = stagePendingUserMessage(input = input)
    metaState.update { current ->
      current.copy(
        pendingUserMessages = current.pendingUserMessages + queuedMessage,
        errorMessage = null,
      )
    }
    Log.d(
      TAG,
      "send accepted sessionId=$sessionId model=${model.name} inputLength=${input.length} pendingCount=${metaState.value.pendingUserMessages.size} messageId=${queuedMessage.message.id}",
    )

    viewModelScope.launch(Dispatchers.Default) {
      playSendSound()
    }

    if (metaState.value.inProgress) {
      requestMergeAndStop(model = model)
      return
    }

    scheduleDispatch(reason = "send accepted")
  }

  fun switchModel(modelId: String) {
    viewModelScope.launch {
      val session = conversationRepository.getSession(sessionId) ?: return@launch
      if (session.activeModelId == modelId) {
        return@launch
      }

      val now = System.currentTimeMillis()
      conversationRepository.updateSession(
        session.copy(activeModelId = modelId, updatedAt = now, lastMessageAt = session.lastMessageAt)
      )
      conversationRepository.appendEvent(
        SessionEvent(
          id = UUID.randomUUID().toString(),
          sessionId = sessionId,
          eventType = SessionEventType.MODEL_SWITCH,
          payloadJson = """{"activeModelId":"${modelId.escapeJson()}"}""",
          createdAt = now,
        )
      )
      metaState.update { current -> current.copy(errorMessage = null) }
      refreshSupplementalState()
    }
  }

  fun pinMessage(message: Message) {
    viewModelScope.launch {
      val session = conversationRepository.getSession(sessionId) ?: return@launch
      val role = roleRepository.getRole(session.roleId) ?: return@launch
      extractMemoriesUseCase.pinMessage(session = session, role = role, message = message)
      refreshSupplementalState()
    }
  }

  fun addManualMemory(content: String, category: MemoryCategory) {
    viewModelScope.launch {
      val session = conversationRepository.getSession(sessionId) ?: return@launch
      val role = roleRepository.getRole(session.roleId) ?: return@launch
      extractMemoriesUseCase.addManualMemory(
        session = session,
        role = role,
        content = content,
        category = category,
      )
      refreshSupplementalState()
    }
  }

  private fun scheduleDispatch(reason: String) {
    val model = latestQueuedModel ?: return
    dispatchJob?.cancel()
    dispatchJob =
      viewModelScope.launch {
        while (true) {
          val delayMs = remainingDispatchDelay()
          if (delayMs <= 0L) {
            break
          }
          Log.d(
            TAG,
            "dispatch paused sessionId=$sessionId reason=$reason delayMs=$delayMs pendingCount=${metaState.value.pendingUserMessages.size}",
          )
          delay(delayMs)
        }

        if (metaState.value.inProgress || metaState.value.pendingUserMessages.isEmpty()) {
          return@launch
        }

        dispatchPendingMessages(model = model)
      }
  }

  private fun requestMergeAndStop(model: Model) {
    if (!metaState.value.inProgress) {
      return
    }

    stopRequested.value = true
    activeDispatchSuperseded = true
    metaState.update { current -> current.copy(errorMessage = null) }
    Log.d(
      TAG,
      "send merge requested sessionId=$sessionId model=${model.name} pendingCount=${metaState.value.pendingUserMessages.size} activeAssistantMessageId=$activeAssistantMessageId",
    )
    model.runtimeHelper.stopResponse(model)
  }

  private fun dispatchPendingMessages(model: Model) {
    val queuedMessages = metaState.value.pendingUserMessages
    if (queuedMessages.isEmpty() || metaState.value.inProgress) {
      return
    }

    val stagedTurn = stageDispatchTurn(userMessages = queuedMessages.map { it.message }, model = model)
    val persistedIds = queuedMessages.filter { it.persisted }.mapTo(mutableSetOf()) { it.message.id }
    val queuedIds = queuedMessages.mapTo(mutableSetOf()) { it.message.id }
    val dispatchStartedAt = SystemClock.elapsedRealtime()

    stopRequested.value = false
    activeAssistantMessageId = stagedTurn.assistantMessage.id
    activeDispatchSuperseded = false
    metaState.update { current ->
      current.copy(
        inProgress = true,
        errorMessage = null,
      )
    }
    Log.d(
      TAG,
      "dispatch starting sessionId=$sessionId model=${model.name} pendingCount=${queuedMessages.size} persistedCount=${persistedIds.size} combinedLength=${stagedTurn.combinedUserInput.length} assistantMessageId=${stagedTurn.assistantMessage.id}",
    )

    viewModelScope.launch(Dispatchers.IO) {
      val pendingMessage =
        sendRoleplayMessageUseCase.enqueuePendingMessage(
          sessionId = sessionId,
          stagedTurn = stagedTurn,
          persistedUserMessageIds = persistedIds,
        )

      if (pendingMessage == null) {
        Log.d(
          TAG,
          "dispatch queue failed after ${SystemClock.elapsedRealtime() - dispatchStartedAt}ms sessionId=$sessionId",
        )
        draft.value = stagedTurn.combinedUserInput
        stopRequested.value = false
        activeAssistantMessageId = null
        metaState.update { current ->
          current.copy(
            pendingUserMessages = current.pendingUserMessages.filterNot { it.message.id in queuedIds },
            inProgress = false,
            errorMessage = "Session no longer exists.",
          )
        }
        return@launch
      }

      metaState.update { current ->
        current.copy(
          pendingUserMessages =
            current.pendingUserMessages.map { queued ->
              if (queued.message.id in queuedIds) {
                queued.copy(persisted = true)
              } else {
                queued
              }
            }
        )
      }
      Log.d(
        TAG,
        "dispatch queued after ${SystemClock.elapsedRealtime() - dispatchStartedAt}ms sessionId=$sessionId assistantMessageId=${pendingMessage.assistantSeed.id}",
      )

      val result =
        sendRoleplayMessageUseCase.completePendingMessage(
          pendingMessage = pendingMessage,
          model = model,
          isStopRequested = { stopRequested.value },
        )
      val superseded = activeDispatchSuperseded

      Log.d(
        TAG,
        "dispatch finished after ${SystemClock.elapsedRealtime() - dispatchStartedAt}ms sessionId=$sessionId interrupted=${result.interrupted} superseded=$superseded error=${result.errorMessage != null}",
      )

      if (superseded && result.assistantMessage != null) {
        conversationRepository.updateMessage(
          result.assistantMessage.copy(
            content = "",
            status = MessageStatus.INTERRUPTED,
            errorMessage = null,
            updatedAt = System.currentTimeMillis(),
          )
        )
      }

      if (!superseded && result.assistantMessage != null && result.assistantMessage.status == MessageStatus.COMPLETED) {
        launch(Dispatchers.Default) {
          playReceiveSound()
        }
      }
      if (!superseded && result.errorMessage != null && !result.interrupted) {
        draft.value = stagedTurn.combinedUserInput
      }

      stopRequested.value = false
      activeAssistantMessageId = null
      activeDispatchSuperseded = false
      metaState.update { current ->
        current.copy(
          pendingUserMessages =
            if (result.interrupted || superseded) {
              current.pendingUserMessages
            } else {
              current.pendingUserMessages.filterNot { it.message.id in queuedIds }
            },
          inProgress = false,
          errorMessage = if (result.interrupted || superseded) null else result.errorMessage,
        )
      }
      refreshSupplementalState()

      if (metaState.value.pendingUserMessages.isNotEmpty()) {
        scheduleDispatch(reason = "pending queue remains after completion")
      }
    }
  }

  private fun refreshSupplementalState() {
    viewModelScope.launch {
      val session = conversationRepository.getSession(sessionId)
      if (session == null) {
        metaState.update { current ->
          current.copy(summary = null, pinnedMemories = emptyList())
        }
        return@launch
      }

      val summary = conversationRepository.getSummary(sessionId)
      val pinnedMemories = loadPinnedMemories(session)
      metaState.update { current ->
        current.copy(summary = summary, pinnedMemories = pinnedMemories)
      }
    }
  }

  private suspend fun loadPinnedMemories(session: Session): List<MemoryItem> {
    return (memoryRepository.listSessionMemories(session.id) + memoryRepository.listRoleMemories(session.roleId))
      .filter { it.pinned }
      .distinctBy { it.normalizedHash }
      .sortedByDescending { it.updatedAt }
      .take(8)
  }

  private fun stagePendingUserMessage(input: String): QueuedUserMessage {
    val now = System.currentTimeMillis()
    val nextSeq = (uiState.value.messages.maxOfOrNull { it.seq } ?: 0) + 1
    val userMessage =
      Message(
        id = UUID.randomUUID().toString(),
        sessionId = sessionId,
        seq = nextSeq,
        side = MessageSide.USER,
        status = MessageStatus.COMPLETED,
        content = input,
        createdAt = now,
        updatedAt = now,
      )
    Log.d(TAG, "queued user draft sessionId=$sessionId seq=$nextSeq messageId=${userMessage.id}")
    return QueuedUserMessage(message = userMessage)
  }

  private fun stageDispatchTurn(userMessages: List<Message>, model: Model): StagedRoleplayTurn {
    val now = System.currentTimeMillis()
    val assistantMessage =
      Message(
        id = UUID.randomUUID().toString(),
        sessionId = sessionId,
        seq = (userMessages.maxOfOrNull { it.seq } ?: 0) + 1,
        side = MessageSide.ASSISTANT,
        status = MessageStatus.STREAMING,
        content = "",
        accelerator = model.getStringConfigValue(key = ConfigKeys.ACCELERATOR, defaultValue = ""),
        parentMessageId = userMessages.lastOrNull()?.id,
        createdAt = now,
        updatedAt = now,
      )
    Log.d(
      TAG,
      "dispatch turn staged sessionId=$sessionId userMessageCount=${userMessages.size} assistantMessageId=${assistantMessage.id}",
    )
    return StagedRoleplayTurn(
      userMessages = userMessages,
      assistantMessage = assistantMessage,
      combinedUserInput = userMessages.joinToString(separator = "\n\n") { it.content.trim() },
    )
  }

  private fun remainingDispatchDelay(): Long {
    val elapsed = SystemClock.elapsedRealtime() - lastDraftEditAtElapsed
    return (SEND_DISPATCH_DELAY_MS - elapsed).coerceAtLeast(0L)
  }

  private fun mergeMessages(messages: List<Message>, queuedMessages: List<QueuedUserMessage>): List<Message> {
    val persistedIds = messages.mapTo(mutableSetOf()) { it.id }
    return (messages + queuedMessages.map { it.message }.filterNot { it.id in persistedIds })
      .filterNot { message ->
        message.side == MessageSide.ASSISTANT &&
          message.status == MessageStatus.INTERRUPTED &&
          message.content.isBlank()
      }
      .sortedWith(compareBy<Message>({ it.seq }, { it.createdAt }, { it.id }))
  }

  private fun String.escapeJson(): String {
    return replace("\\", "\\\\").replace("\"", "\\\"")
  }

  private fun playSendSound() {
    if (!dataStoreRepository.areMessageSoundsEnabled()) {
      return
    }
    RoleplaySoundEffectPlayer.playSend(appContext)
  }

  private fun playReceiveSound() {
    if (!dataStoreRepository.areMessageSoundsEnabled()) {
      return
    }
    RoleplaySoundEffectPlayer.playReceive(appContext)
  }
}
